homeassistant/components/switchbot_cloud/fan.py

```diff
@@ -0,0 +1,197 @@
+        preset_mode: str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Turn on the fan."""
+        response: dict | None = await self._api.get_status(self.unique_id)
+        assert response is not None
+
+        power: str | None = response.get("power")
+        fan_speed: int | None = response.get("fanSpeed")
+        mode: str | None = response.get("mode")
+        if self._attr_is_on is False and (power and "off" in power):
+            await self.send_api_command(CommonCommands.ON)
+            self._attr_is_on = True
+        elif self._attr_is_on is False and (power and "on" in power):
+            self._attr_is_on = True
+        else:
+            pass
+        if self.preset_mode is None:
+            self.preset_mode = BatteryCirculatorFanMode.DIRECT.value
+        # if mode exist, do something
+        if mode is not None:
+            # if current mode != preset_mode, send set mode command again
+            if mode != self.preset_mode:
+                await self.send_api_command(
+                    command=BatteryCirculatorFanCommands.SET_WIND_MODE,
+                    parameters=self.preset_mode,
+                )
+        # if mode is None, set as default
+        else:
+            self.preset_mode = BatteryCirculatorFanMode.DIRECT.value
+            await self.send_api_command(
+                command=BatteryCirculatorFanCommands.SET_WIND_MODE,
+                parameters=self.preset_mode,
+            )
+
+        # if fanspeed exist, do something
+        if (
+            fan_speed is not None
+            and fan_speed != self.percentage
+            and self.preset_mode in BatteryCirculatorFanMode.DIRECT.value
+        ):
+            await self.send_api_command(
+                command=BatteryCirculatorFanCommands.SET_WIND_SPEED,
+                parameters=str(percentage),
+            )
+        else:
+            self.percentage = 0
+        self.async_write_ha_state()
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn off the fan."""
+
+        response: dict | None = await self._api.get_status(self.unique_id)
+        assert response is not None
+        power: str | None = response.get("power")
+        fan_speed: int | None = response.get("fanSpeed")
+        mode: str | None = response.get("mode")
+
+        # update self.percentage and self.preset_mode
+        if fan_speed is not None:
+            self.percentage = fan_speed
+        if mode is not None:
+            self.preset_mode = mode
+            if mode not in BatteryCirculatorFanMode.DIRECT.value:
+                self.percentage = 0
+
+        if self._attr_is_on is True and (power and "on" in power):
+            await self.send_api_command(CommonCommands.OFF)
+            self._attr_is_on = False
+        elif self._attr_is_on is True and (power and "off" in power):
+            self._attr_is_on = False
+        else:
+            pass
+        self.async_write_ha_state()
```

Comment:
So currently we update the state and then write state, but what if the update doesn't go through on your side? Maybe the bridge is out of range or the fan doesn't work for some reason? Wouldn't it be better to fetch new state?